File: social.js
// ════════════════════════════════════════════════════════════════
// Product Scout — Social Discovery view
// Talks to /api/social/* (Reddit now; Google Trends / others later).
// ════════════════════════════════════════════════════════════════

const sentClass = label =>
  (label || '').includes('positive') ? 'sent-pos' :
  (label || '').includes('negative') ? 'sent-neg' : 'sent-neu';

const esc = s => { const d = document.createElement('div'); d.textContent = s == null ? '' : s; return d.innerHTML; };

let SOCIAL_OV = null;
let socialLoaded = false;

// ───────────────────────── view switching ─────────────────────────
document.querySelectorAll('.nav-item').forEach(btn => {
  btn.addEventListener('click', () => {
    const view = btn.dataset.view;
    document.querySelectorAll('.nav-item').forEach(b => b.classList.toggle('active', b === btn));
    document.querySelectorAll('.view').forEach(v => v.classList.remove('active'));
    document.getElementById('view-' + view).classList.add('active');
    // year/HS controls only belong to the Trade view
    document.getElementById('tradeControls').style.display = view === 'trade' ? 'flex' : 'none';
    if (view === 'trade') {
      window.dispatchEvent(new Event('resize'));   // refit the map
    } else if (view === 'social' && !socialLoaded) {
      loadSocialOverview();
    } else if (view === 'fusion' && window.loadFusion && !window.fusionLoaded) {
      window.loadFusion();
    }
  });
});

// ───────────────────────── overview ─────────────────────────
async function loadSocialOverview() {
  socialLoaded = true;
  try {
    SOCIAL_OV = await fetch('/api/social/overview').then(r => r.json());
  } catch (e) { console.error(e); return; }
  // example chips
  document.getElementById('exampleChips').innerHTML =
    SOCIAL_OV.example_queries.map(q => `<div class="chip" data-q="${esc(q)}">${esc(q)}</div>`).join('');
  document.querySelectorAll('#exampleChips .chip').forEach(c =>
    c.addEventListener('click', () => runSearch(c.dataset.q)));

  renderOverview();
}

function pill(name, label, cls) {
  return `<span class="product-pill"><span class="dot ${sentClass(label)}"></span>${esc(name)}
            <span class="${sentClass(label)}" style="font-size:10px">${label && label !== 'n/a' ? label : ''}</span></span>`;
}

const ANGLE_CLASS = { "Unmet need": "whitespace", "Room to differentiate": "differentiate", "Validated demand": "validated" };

function renderInsights(ins) {
  if (!ins || !ins.opportunities || !ins.opportunities.length) return '';
  const h = ins.headline;
  const cards = ins.opportunities.map(o => `
    <div class="opp-launch">
      <span class="tag-angle ${ANGLE_CLASS[o.angle] || 'differentiate'}">${esc(o.angle)}</span>
      <h4>${esc(o.category)}</h4>
      <div class="meta">${o.n_posts} conversations</div>
      <div class="why">${esc(o.rationale)}</div>
      <div class="incumbents">Incumbents to beat: <strong>${o.incumbents.map(esc).join(', ') || '—'}</strong></div>
      <button class="cta-btn" data-q="${esc(o.category)}">Explore demand →</button>
    </div>`).join('');
  return `
    <div class="insights-panel">
      <div class="insights-head">
        <div>
          <h3>Launch Opportunities</h3>
          <p>Trending categories ranked by conversation volume and unmet need — where social demand signals a clear opening to launch new products.</p>
        </div>
        <div class="insights-stats">
          <div class="stat"><div class="v">${(h.n_posts || 0).toLocaleString()}</div><div class="l">posts analyzed</div></div>
          <div class="stat"><div class="v">${h.n_categories || 0}</div><div class="l">categories</div></div>
        </div>
      </div>
      <div class="opp-cards">${cards}</div>
    </div>`;
}

function renderOverview() {
  const ov = SOCIAL_OV;
  const card = c => `
    <div class="ov-card" data-q="${esc(c.category)}">
      <h4>${esc(c.category)}</h4>
      <div class="ov-meta">
        <span>${c.n_posts} posts</span>
      </div>
    </div>`;
  const TOP_N = 8;
  const topCards = ov.categories.slice(0, TOP_N).map(card).join('');
  const restCount = Math.max(0, ov.categories.length - TOP_N);
  const restCards = ov.categories.slice(TOP_N).map(card).join('');
  const catCards = `<div class="ov-grid">${topCards}</div>` + (restCount ? `
    <div class="ov-grid" id="ovGridRest" style="display:none;margin-top:14px">${restCards}</div>
    <button class="home-btn" id="ovMoreBtn" style="margin-top:14px">▸ Show ${restCount} more categories</button>` : '');

  document.getElementById('socialOverview').innerHTML = `
    <div class="section-title" style="margin-top:8px"><div class="dot" style="background:var(--accent)"></div> Browse by product category <span style="font-size:11px;color:var(--text2);font-weight:400">— click a category for Amazon reviews & search trends</span></div>
    ${catCards}

    ${renderInsights(ov.insights)}

    <div class="section-title" style="margin-top:28px"><div class="dot" style="background:var(--accent)"></div> Most-discussed products</div>
    <div class="ent-list">${ov.top_brands.map(b => pill(b.entity, label(b.avg_sentiment))).join('')}</div>

    <div class="section-title" style="margin-top:24px"><div class="dot" style="background:var(--accent)"></div> Trending ingredients</div>
    <div class="ent-list">${ov.top_ingredients.map(i => pill(i.entity, label(i.avg_sentiment))).join('')}</div>`;

  // category cards & launch-opportunity CTAs open the category detail page;
  // free-text pills/chips still run a search.
  document.querySelectorAll('#socialOverview .ov-card, #socialOverview .cta-btn').forEach(c =>
    c.addEventListener('click', () => window.openCategory ? window.openCategory(c.dataset.q) : runSearch(c.dataset.q)));

  // show more / fewer categories
  const moreBtn = document.getElementById('ovMoreBtn');
  if (moreBtn) {
    const rest = document.getElementById('ovGridRest');
    const n = rest.querySelectorAll('.ov-card').length;
    moreBtn.addEventListener('click', () => {
      const open = rest.style.display === 'none';
      rest.style.display = open ? 'grid' : 'none';
      moreBtn.textContent = open ? '▾ Show fewer categories' : `▸ Show ${n} more categories`;
    });
  }
}

// return to the categories home view
function goHome() {
  document.getElementById('socialSearch').value = '';
  document.getElementById('socialResults').innerHTML = '';
  const cat = document.getElementById('socialCategory');
  if (cat) { cat.style.display = 'none'; cat.innerHTML = ''; }
  const ss = document.getElementById('socialSearchSection');   // restore title + search
  if (ss) ss.style.display = '';
  document.getElementById('socialOverview').style.display = 'block';
  window.scrollTo({ top: 0, behavior: 'smooth' });
}
window.goHome = goHome;

// client-side sentiment label (mirror of server thresholds) for overview numbers
function label(v) {
  if (v == null) return 'n/a';
  if (v >= 0.4) return 'very positive';
  if (v >= 0.1) return 'positive';
  if (v > -0.1) return 'neutral';
  if (v > -0.4) return 'negative';
  return 'very negative';
}

// ───────────────────────── search ─────────────────────────
async function runSearch(q) {
  const input = document.getElementById('socialSearch');
  input.value = q;
  const resBox = document.getElementById('socialResults');
  const ovBox = document.getElementById('socialOverview');
  resBox.innerHTML = `<div style="text-align:center;color:var(--text2);padding:30px">Analyzing posts for “${esc(q)}”…</div>`;
  ovBox.style.display = 'none';
  let data;
  try {
    data = await fetch('/api/social/search?q=' + encodeURIComponent(q) + '&limit=25').then(r => r.json());
  } catch (e) { resBox.innerHTML = `<div style="color:var(--negative)">Search failed: ${esc(String(e))}</div>`; return; }
  renderResults(data);
}

function tagRow(p) {
  const t = [];
  p.categories.slice(0, 4).forEach(c => t.push(`<span class="tag cat">${esc(c)}</span>`));
  p.brands.slice(0, 4).forEach(b => t.push(`<span class="tag brand">${esc(b)}</span>`));
  p.ingredients.slice(0, 4).forEach(i => t.push(`<span class="tag ing">${esc(i)}</span>`));
  return t.join('');
}

function sentBadge(val, label, prefix) {
  if (val == null) return '';
  return `<span class="sent-badge ${sentClass(label)}">${prefix} ${label} (${val >= 0 ? '+' : ''}${val.toFixed(2)})</span>`;
}

function renderResults(data) {
  const s = data.summary;
  const resolvedBits = Object.entries(data.resolved).flatMap(([t, names]) =>
    names.map(n => `<span class="tag ${t === 'category' ? 'cat' : t === 'brand' ? 'brand' : 'ing'}">${esc(n)}</span>`)).join(' ');

  const backBtn = `<button class="home-btn" id="socialBackBtn">← Back to categories</button>`;

  let html = backBtn;
  if (!data.results.length) {
    html += `<div class="result-card">No posts matched “${esc(data.query)}”. Try a broader term like “acne” or “moisturizer”.</div>`;
    const box = document.getElementById('socialResults');
    box.innerHTML = html;
    box.querySelector('#socialBackBtn').addEventListener('click', goHome);
    return;
  }

  // summary strip
  html += `
    <div class="summary-strip">
      <div>
        <div class="blk-label">Overall sentiment</div>
        <div class="big-sent ${sentClass(s.sentiment_label)}">${s.avg_sentiment >= 0 ? '+' : ''}${(s.avg_sentiment ?? 0).toFixed(2)}</div>
        <div class="${sentClass(s.sentiment_label)}" style="font-size:12px">${s.sentiment_label} · ${s.n_results} posts${resolvedBits ? ' · matched ' + resolvedBits : ''}</div>
      </div>
      <div>
        <div class="blk-label">Top products mentioned</div>
        <div>${s.top_products.length ? s.top_products.map(p => pill(`${p.name} ·${p.mentions_in_results}`, p.sentiment_label)).join('') : '<span style="color:var(--text2);font-size:12px">none detected</span>'}</div>
      </div>
      <div>
        <div class="blk-label">Top ingredients mentioned</div>
        <div>${s.top_ingredients.length ? s.top_ingredients.map(p => pill(`${p.name} ·${p.mentions_in_results}`, p.sentiment_label)).join('') : '<span style="color:var(--text2);font-size:12px">none detected</span>'}</div>
      </div>
    </div>
    <div class="section-title"><div class="dot" style="background:var(--accent)"></div> ${data.results.length} relevant posts</div>`;

  // result cards
  html += data.results.map(p => {
    const link = p.permalink ? `<a href="${esc(p.permalink)}" target="_blank" rel="noopener">${esc(p.title)}</a>` : esc(p.title);
    const comment = p.top_comment ? `
      <div class="rc-comment">“${esc(p.top_comment.body)}”
        <div class="cm-meta">▲ ${p.top_comment.ups} · <span class="${sentClass(p.top_comment.label)}">${p.top_comment.label}</span></div>
      </div>` : '';
    return `
      <div class="result-card">
        <div class="rc-head">
          <div>
            <div class="rc-title">${link}</div>
            <div class="rc-sub">
              <span class="sub-badge">r/${esc(p.subreddit)}</span>
              <span>${p.n_comments} comments</span>
              ${sentBadge(p.post_sentiment, p.post_sentiment_label, 'post:')}
              ${sentBadge(p.discussion_sentiment, p.discussion_label, 'discussion:')}
            </div>
          </div>
        </div>
        ${p.snippet ? `<div class="rc-snippet">${esc(p.snippet)}</div>` : ''}
        <div class="rc-tags">${tagRow(p)}</div>
        ${comment}
      </div>`;
  }).join('');

  const box = document.getElementById('socialResults');
  box.innerHTML = html;
  box.querySelector('#socialBackBtn').addEventListener('click', goHome);
}

// reset to overview when search cleared
document.getElementById('socialSearchBtn').addEventListener('click', () => {
  const q = document.getElementById('socialSearch').value.trim();
  if (q) runSearch(q);
});
document.getElementById('socialSearch').addEventListener('keydown', e => {
  if (e.key === 'Enter') { const q = e.target.value.trim(); if (q) runSearch(q); }
  if (e.key === 'Escape') {
    e.target.value = '';
    document.getElementById('socialResults').innerHTML = '';
    document.getElementById('socialOverview').style.display = 'block';
  }
});

File: social.py
"""
Social discovery query layer for Product Scout.

Loads the processed Reddit data from PostgreSQL and answers
e-commerce-style queries like "oily skin remover" — resolving them to product
categories / brands / ingredients, then returning the most relevant posts with
sentiment scoring and the products mentioned in them.
"""

import os
import json
import re
import urllib.parse
from collections import Counter, defaultdict
from pathlib import Path
from sqlalchemy import create_engine, text
from dotenv import load_dotenv

from social_ingest import build_matchers, extract

# -------------------------------------------------------------------------
# DATABASE CONNECTION SETUP (Safely handles special characters)
# -------------------------------------------------------------------------
load_dotenv()

db_user = os.getenv("DB_USER", "postgres")
db_password = urllib.parse.quote_plus(os.getenv("DB_PASSWORD", ""))
db_host = os.getenv("DB_HOST", "localhost")
db_port = os.getenv("DB_PORT", "5432")
db_name = os.getenv("DB_NAME", "capstone_db")

DATABASE_URL = os.getenv("DATABASE_URL") or f"postgresql://{db_user}:{db_password}@{db_host}:{db_port}/{db_name}"
engine = create_engine(DATABASE_URL)

STOPWORDS = {
    "the", "a", "an", "to", "for", "of", "and", "or", "my", "me", "i", "is", "in",
    "on", "with", "best", "good", "any", "what", "which", "how", "do", "does",
    "remover", "remove", "removal", "product", "products", "recommend",
    "recommendation", "recommendations", "help", "need", "want", "looking", "find",
    "skin", "skincare", "face", "routine", "use", "using", "get", "rid",
}


def _label(v):
    if v is None:
        return "n/a"
    if v >= 0.4:
        return "very positive"
    if v >= 0.1:
        return "positive"
    if v > -0.1:
        return "neutral"
    if v > -0.4:
        return "negative"
    return "very negative"


class SocialData:
    def __init__(self):
        """
        DATABASE MIGRATION: Replaced original flat JSON file loads with direct PostgreSQL queries.
        Casts database numeric types to primitive floats/ints to avoid Decimal type collisions downstream.
        """
        print("📥 Initializing SocialData: Fetching analytical layers from PostgreSQL...")

        with engine.connect() as conn:
            # 1. Load posts from social_posts table
            post_rows = conn.execute(text("SELECT * FROM social_posts")).mappings().all()
            self.posts = []
            for r in post_rows:
                p = dict(r)
                if isinstance(p.get("categories"), str): p["categories"] = json.loads(p["categories"])
                if isinstance(p.get("brands"), str): p["brands"] = json.loads(p["brands"])
                if isinstance(p.get("ingredients"), str): p["ingredients"] = json.loads(p["ingredients"])
                if isinstance(p.get("entities"), str): p["entities"] = json.loads(p["entities"])
                if isinstance(p.get("top_comments"), str): p["top_comments"] = json.loads(p["top_comments"])
                
                # Safe Type Casting
                p["post_sentiment"] = float(p["post_sentiment"]) if p["post_sentiment"] is not None else 0.0
                p["discussion_sentiment"] = float(p["discussion_sentiment"]) if p["discussion_sentiment"] is not None else 0.0
                p["n_comments"] = int(p["n_comments"]) if p["n_comments"] is not None else 0
                self.posts.append(p)

            # 2. Load global entities rollup summary
            entity_rows = conn.execute(text("SELECT * FROM social_entities")).mappings().all()
            self.entities = []
            for r in entity_rows:
                e = dict(r)
                e["mentions"] = int(e["mentions"]) if e["mentions"] is not None else 0
                e["n_posts"] = int(e["n_posts"]) if e["n_posts"] is not None else 0
                e["avg_sentiment"] = float(e["avg_sentiment"]) if e["avg_sentiment"] is not None else 0.0
                self.entities.append(e)

            # 3. Load global categories rollup summary
            cat_rows = conn.execute(text("SELECT * FROM social_categories")).mappings().all()
            self.categories = []
            for r in cat_rows:
                c = dict(r)
                if isinstance(c.get("top_products"), str): c["top_products"] = json.loads(c["top_products"])
                if isinstance(c.get("top_ingredients"), str): c["top_ingredients"] = json.loads(c["top_ingredients"])
                
                c["n_posts"] = int(c["n_posts"]) if c["n_posts"] is not None else 0
                c["avg_sentiment"] = float(c["avg_sentiment"]) if c["avg_sentiment"] is not None else 0.0
                self.categories.append(c)

            # 4. Load metadata snapshot
            meta_row = conn.execute(text("SELECT * FROM social_meta LIMIT 1")).mappings().first()
            if meta_row:
                m = dict(meta_row)
                if isinstance(m.get("sources"), str): m["sources"] = json.loads(m["sources"])
                if isinstance(m.get("subreddits"), str): m["subreddits"] = json.loads(m["subreddits"])
                
                m["n_posts"] = int(m["n_posts"]) if m["n_posts"] is not None else 0
                m["n_comments"] = int(m["n_comments"]) if m["n_comments"] is not None else 0
                m["avg_post_sentiment"] = float(m["avg_post_sentiment"]) if m["avg_post_sentiment"] is not None else 0.0
                self.meta = m
            else:
                self.meta = {"sources": ["reddit"], "subreddits": [], "n_posts": 0, "n_comments": 0}

        # PERFECT MATCH: Retaining original underscore naming contracts used by downstream methods
        self.matchers = build_matchers()
        self._by_id = {p["id"]: p for p in self.posts}
        self._ent_sent = {(e["entity_type"], e["entity"]): e for e in self.entities}

        print(f"✅ Successfully cached {len(self.posts)} processed posts from database into memory.")

    # ---------- overview for the discovery landing ----------
    def overview(self):
        def by_type(t, n):
            return [e for e in self.entities if e["entity_type"] == t][:n]
        return {
            "meta": self.meta,
            "insights": self.launch_insights(),
            "categories": self.categories,
            "top_brands": by_type("brand", 12),
            "top_ingredients": by_type("ingredient", 12),
            "example_queries": [
                "oily skin remover", "dark spots and hyperpigmentation",
                "gentle cleanser for sensitive skin", "retinol for wrinkles",
                "korean sunscreen", "fungal acne safe moisturizer",
            ],
        }

    # ---------- launch-opportunity insights ----------
    def launch_insights(self, n=4):
        """Trending categories + a product-launch opportunity read.

        Signal = demand (conversation volume) blended with an unmet-need bonus
        (lower consumer sentiment in a high-volume category = white space).
        Timestamps span only ~2 weeks, so we deliberately rank by demand, not a
        fabricated time-trend.
        """
        cats = self.categories
        if not cats:
            return {"headline": {}, "opportunities": []}
        max_posts = max(c["n_posts"] for c in cats) or 1

        scored = []
        for c in cats:
            s = c["avg_sentiment"] or 0.0
            demand = c["n_posts"] / max_posts
            unmet = max(0.0, 0.25 - s)            # bonus when satisfaction is low
            scored.append((demand * (1 + unmet * 2.5), c))
        scored.sort(key=lambda x: x[0], reverse=True)

        opps = []
        for score, c in scored[:n]:
            s = c["avg_sentiment"] or 0.0
            if s >= 0.25:
                angle = "Validated demand"
                why = "shoppers are enthusiastic — a proven market to enter with a strong product"
            elif s >= 0.1:
                angle = "Room to differentiate"
                why = "high interest but only mild satisfaction — a better formulation can stand out"
            else:
                angle = "Unmet need"
                why = "lots of discussion but low satisfaction — clear white space for a product that delivers"
            opps.append({
                "category": c["category"],
                "n_posts": c["n_posts"],
                "avg_sentiment": c["avg_sentiment"],
                "sentiment_label": _label(c["avg_sentiment"]),
                "angle": angle,
                "rationale": f"{c['n_posts']} conversations — {why}.",
                "incumbents": c["top_products"][:3],
                "opportunity_score": round(score, 3),
            })

        m = self.meta
        return {
            "headline": {
                "n_posts": m.get("n_posts"),
                "n_comments": m.get("n_comments"),
                "n_categories": len(cats),
                "overall_sentiment": m.get("avg_post_sentiment"),
                "overall_label": _label(m.get("avg_post_sentiment")),
                "top_category": cats[0]["category"] if cats else None,
            },
            "opportunities": opps,
        }

    # ---------- search ----------
    def search(self, query, limit=25):
        q = (query or "").strip()
        if not q:
            return {"query": q, "resolved": {}, "summary": {}, "results": []}

        targets = extract(q, self.matchers)            # {etype: Counter}
        target_set = {(et, name) for et, c in targets.items() for name in c}
        tokens = [t for t in re.findall(r"[a-z0-9]+", q.lower())
                  if t not in STOPWORDS and len(t) > 2]

        scored = []
        for p in self.posts:
            score = 0.0
            # entity overlap (strong signal)
            for et, ents in p["entities"].items():
                for e in ents:
                    if (et, e["name"]) in target_set:
                        w = 5 if et != "category" else 4
                        score += w + min(e["mentions"], 4) * 0.5
            # free-text relevance
            title_low = p["title"].lower()
            blob = p["search_blob"]
            for t in tokens:
                if t in title_low:
                    score += 3
                hits = blob.count(t)
                if hits:
                    score += min(hits, 5) * 0.6
            if score > 0:
                scored.append((score, p))

        scored.sort(key=lambda x: (x[0], x[1]["n_comments"]), reverse=True)
        top = scored[:limit]

        # aggregate "potential products" across the matched posts
        brand_ct, ing_ct = Counter(), Counter()
        sent_vals = []
        for _, p in top:
            brand_ct.update(p["brands"])
            ing_ct.update(p["ingredients"])
            sent_vals.append(p["post_sentiment"])

        def products(counter, etype):
            out = []
            for name, n in counter.most_common(8):
                es = self._ent_sent.get((etype, name), {})
                out.append({"name": name, "mentions_in_results": n,
                            "avg_sentiment": es.get("avg_sentiment"),
                            "sentiment_label": _label(es.get("avg_sentiment"))})
            return out

        avg_sent = round(sum(sent_vals) / len(sent_vals), 4) if sent_vals else None

        return {
            "query": q,
            "resolved": {et: list(c.keys()) for et, c in targets.items()},
            "summary": {
                "n_results": len(top),
                "avg_sentiment": avg_sent,
                "sentiment_label": _label(avg_sent),
                "top_products": products(brand_ct, "brand"),
                "top_ingredients": products(ing_ct, "ingredient"),
            },
            "results": [self._result_card(p, score) for score, p in top],
        }

    # ---------- compact helpers for the Claude agent ----------
    def agent_search(self, query, limit=8):
        """Trimmed search result suitable for an LLM tool response."""
        r = self.search(query, limit=limit)
        s = r["summary"]
        return {
            "query": r["query"],
            "resolved_to": r["resolved"],
            "n_matching_posts": s.get("n_results", 0),
            "overall_sentiment": s.get("avg_sentiment"),
            "overall_sentiment_label": s.get("sentiment_label"),
            "top_products": [{"name": p["name"], "mentions": p["mentions_in_results"],
                              "sentiment": p["avg_sentiment"], "label": p["sentiment_label"]}
                             for p in s.get("top_products", [])],
            "top_ingredients": [{"name": p["name"], "mentions": p["mentions_in_results"],
                                 "sentiment": p["avg_sentiment"], "label": p["sentiment_label"]}
                                for p in s.get("top_ingredients", [])],
            "sample_posts": [{"title": c["title"], "subreddit": c["subreddit"],
                              "post_sentiment": c["post_sentiment"],
                              "discussion_sentiment": c["discussion_sentiment"],
                              "brands": c["brands"][:4], "n_comments": c["n_comments"]}
                             for c in r["results"][:6]],
        }

    def agent_overview(self):
        return {
            "sources": self.meta.get("sources"),
            "n_posts": self.meta.get("n_posts"),
            "n_comments": self.meta.get("n_comments"),
            "categories": [{"category": c["category"], "n_posts": c["n_posts"],
                            "avg_sentiment": c["avg_sentiment"],
                            "top_products": c["top_products"][:4]} for c in self.categories],
            "top_brands": [{"name": e["entity"], "mentions": e["mentions"],
                            "n_posts": e["n_posts"], "sentiment": e["avg_sentiment"]}
                           for e in self.entities if e["entity_type"] == "brand"][:12],
            "top_ingredients": [{"name": e["entity"], "mentions": e["mentions"],
                                 "n_posts": e["n_posts"], "sentiment": e["avg_sentiment"]}
                                for e in self.entities if e["entity_type"] == "ingredient"][:12],
        }

    def product_sentiment(self, name):
        """Fuzzy-find a brand / ingredient / category and return its sentiment rollup."""
        key = (name or "").strip().lower()
        if not key:
            return {"error": "empty name"}
        # exact then substring match across all entities
        cand = [e for e in self.entities if e["entity"].lower() == key] or \
               [e for e in self.entities if key in e["entity"].lower()]
        if not cand:
            return {"error": f"'{name}' not found among tracked products/ingredients/categories."}
        e = max(cand, key=lambda x: x["mentions"])
        return {"name": e["entity"], "type": e["entity_type"], "mentions": e["mentions"],
                "n_posts": e["n_posts"], "avg_sentiment": e["avg_sentiment"],
                "sentiment_label": _label(e["avg_sentiment"])}

    def _result_card(self, p, score):
        top_comment = None
        if p["top_comments"]:
            tc = p["top_comments"][0]
            top_comment = {"body": tc["body"], "ups": tc["ups"],
                           "sentiment": tc["sentiment"], "label": _label(tc["sentiment"])}
        return {
            "id": p["id"], "source": p["source"], "subreddit": p["subreddit"],
            "title": p["title"],
            "snippet": (p["content"][:240] + "…") if len(p["content"]) > 240 else p["content"],
            "permalink": ("https://reddit.com" + p["permalink"]) if p.get("permalink", "").startswith("/") else p.get("permalink"),
            "post_sentiment": p["post_sentiment"],
            "post_sentiment_label": _label(p["post_sentiment"]),
            "discussion_sentiment": p["discussion_sentiment"],
            "discussion_label": _label(p["discussion_sentiment"]),
            "n_comments": p["n_comments"],
            "categories": p["categories"],
            "brands": p["brands"],
            "ingredients": p["ingredients"],
            "top_comment": top_comment,
            "relevance": round(score, 1),
        }

    # ---------- category page bundle (Amazon + Trends + Reddit) ----------
    def category_page(self, name):
        from amazon import get_amazon
        from trends import get_trends
        amazon = get_amazon().category(name)
        reddit = self.search(name, limit=6)
        cat = next((c for c in self.categories if c["category"] == name), None)
        return {
            "category": name,
            "social_summary": {
                "n_posts": cat["n_posts"] if cat else 0,
                "avg_sentiment": cat["avg_sentiment"] if cat else None,
                "sentiment_label": _label(cat["avg_sentiment"]) if cat else "n/a",
                "top_products": cat["top_products"] if cat else [],
                "top_ingredients": cat["top_ingredients"] if cat else [],
            },
            "amazon": amazon,
            "trends": get_trends().category_block(name),
            "reddit": {
                "summary": reddit["summary"],
                "posts": reddit["results"][:5],
            },
        }

_social = None


def get_social():
    global _social
    if _social is None:
        _social = SocialData()
    return _social